Sources/FoundationEssentials/URL/URITemplate_Expression.swift

```diff
@@ -0,0 +1,272 @@
+internal import RegexBuilder
+#if canImport(CollectionsInternal)
+internal import CollectionsInternal
+#elseif canImport(OrderedCollections)
+internal import OrderedCollections
+#elseif canImport(_FoundationCollections)
+internal import _FoundationCollections
+#endif
+
+extension URL.Template {
+    struct Expression: Sendable, Hashable {
+        var `operator`: Operator?
+        var elements: [Element]
+
+        struct Element: Sendable, Hashable {
+            var name: URL.Template.VariableName
+            var maximumLength: Int?
+            var explode: Bool
+        }
+
+        enum Operator: String, Sendable, Hashable {
+            /// `+`   Reserved character strings;
+            case reserved = "+"
+            /// `#`   Fragment identifiers prefixed by "#";
+            case fragment = "#"
+            /// `.`   Name labels or extensions prefixed by ".";
+            case nameLabel = "."
+            /// `/`   Path segments prefixed by "/";
+            case pathSegment = "/"
+            /// `;`   Path parameter name or name=value pairs prefixed by ";";
+            case pathParameter = ";"
+            /// `?`   Query component beginning with "?" and consisting of
+            /// name=value pairs separated by "&"; and,
+            case queryComponent = "?"
+            /// `&`   Continuation of query-style &name=value pairs within
+            /// a literal query component.
+            case continuation = "&"
+        }
+    }
+}
+
+private struct InvalidTemplateExpression: Swift.Error {
+    var text: String
+}
+
+extension Substring {
+    mutating func popPrefixMatch<Output>(_ regex: Regex<Output>) throws -> Regex<Output>.Match? {
```

Comment:
If possible I'd like to mark these kinds of helpers as `fileprivate`, to make it clear where they are used. This makes it far easier to refactor in the future.